src/storage.ts

```diff
@@ -0,0 +1,144 @@
+ *  limitations under the License.
+ */
+
+import { openDB, DBSchema, IDBPDatabase } from 'idb';
+
+const mainDBName = 'appDB';
+const mainImageName = 'mainImage';
+
+export type FileUniqueID = string;  // generated by the storage layer
+export type Timestamp = number;
+
+export type FileMetadata = {
+  filename: string;
+  mimeType: string;
+  title?: string;
+}
+
+export type FileRecord = {
+  id: FileUniqueID;
+  metadata: FileMetadata;
+  added: Timestamp;
+  blobModified: Timestamp;
+  metadataModified: Timestamp;
+};
+
+interface LlamaDB extends DBSchema {
+  blob: {
+    key: string;
+    value: Blob;
+  };
+  metadata: {
+    key: string;
+    value: FileRecord;
+  };
+}
+
+export default class LlamaStorage {
+  private db: IDBPDatabase<LlamaDB>;
+
+  private constructor(db: IDBPDatabase<LlamaDB>) {
+    console.log('Database initialized');
+    this.db = db;
+  }
+
+  static async create() {
+    return new LlamaStorage(await openDB(mainDBName, undefined /* version */, {
+      upgrade(db, oldVersion, newVersion, transaction) {
+        console.log(`DB update request`);
+        db.createObjectStore('blob');
+        db.createObjectStore('metadata');
+      },
+      // blocked()
+      // blocking()
+      // terminated()
+    }));
+  }
+
+  async add(file: Blob, metadata: FileMetadata): Promise<FileRecord> {
+    const id: FileUniqueID = mainImageName /* TODO: generate unique IDs */;
+    const now = Date.now();
+    const record: FileRecord = {
+      id: id,
+      metadata: metadata,
+      added: now,
+      blobModified: now,
+      metadataModified: now,
+    };
+
+    // TODO: it would be nice if the caller could access the FileRecord
+    // immediately, but then still indicate when the data finished
+    // being stored.
+    await Promise.allSettled([
+      this.db.put('blob', file, id),
+      this.db.put('metadata', record, id),
+    ]);
+    console.log(`successfully added '${id}'`);
+    return record;
+  }
+
+  async list(): Promise<FileRecord[]> {
+    // TODO: https://github.com/jakearchibald/idb#async-iterators
+    throw 'list() is unimplemented';
+  }
+
+  async getFile(id: FileUniqueID): Promise<Blob> {
+    const blob = await this.db.get('blob', id);
+    if (!blob) throw `no blob found for id '${id}'`;
+    console.log(`successfully retrieved blob for '${id}'`);
+    return blob;
+  }
+
+  async get(id: FileUniqueID): Promise<FileRecord> {
+    const record = await this.db.get('metadata', id);
+    if (!record) throw `no record found for id '${id}'`;
+    console.log(`successfully retrieved metadata for '${id}'`);
+    return record;
+  }
+
+  // At least one of |file| and |metadata| must be provided.
+  async update(id: FileUniqueID, file?: Blob, metadata?: FileMetadata): Promise<FileRecord> {
+    if (!file && !metadata) throw 'neither |file| nor |metadata| was provided';
+
+    // don't "await" yet, so that we can kick off the blob "put" (if applicable)
+    // first.
```

Comment:
This seems like a micro-optimization? Would be good to call it out as such, particularly because it's not explained why this makes a significant difference.

Reply:
yeaahhh, I'm leaning more and more towards "revert [`6a9dde7` (#41)](https://github.com/GoogleChromeLabs/llaminator/pull/41/commits/6a9dde7f059407401a4b90f1ed252e983dc54b94)"

Reply:
(reverted)